tests/functions/test_expand_grid.py

```diff
@@ -0,0 +1,318 @@
+
+
+def test_computation_output_2():
+    """Test output if entry contains only dataframes/series"""
+    data = {
+        "df": pd.DataFrame({
+            "x": range(1, 6),
+            "y": [5, 4, 3, 2, 1]
+        }),
+        "df1": pd.DataFrame({
+            "x": range(4, 7),
+            "y": [6, 5, 4]
+        }),
+    }
+
+    expected = pd.DataFrame({
+        "df_x": [1, 1, 1, 2, 2, 2, 3, 3, 3, 4, 4, 4, 5, 5, 5],
+        "df_y": [5, 5, 5, 4, 4, 4, 3, 3, 3, 2, 2, 2, 1, 1, 1],
+        "df1_x": [4, 5, 6, 4, 5, 6, 4, 5, 6, 4, 5, 6, 4, 5, 6],
+        "df1_y": [6, 5, 4, 6, 5, 4, 6, 5, 4, 6, 5, 4, 6, 5, 4],
+    })
+
+    assert_frame_equal(expand_grid(others=data), expected)
+
+
+def test_computation_output_3():
+    """Test mix of dataframes and lists"""
+    data = {
+        "df": pd.DataFrame({
+            "x": range(1, 3),
+            "y": [2, 1]
+        }),
+        "z": range(1, 4),
+    }
+    expected = pd.DataFrame({
+        "df_x": [1, 1, 1, 2, 2, 2],
+        "df_y": [2, 2, 2, 1, 1, 1],
+        "z": [1, 2, 3, 1, 2, 3],
+    })
+    assert_frame_equal(expand_grid(others=data), expected)
+
+
+def test_computation_output_4():
+    """ Test output from list of strings"""
+    data = {"l1": list(ascii_lowercase[:3]), "l2": list(ascii_uppercase[:3])}
+    expected = pd.DataFrame({
+        "l1": ["a", "a", "a", "b", "b", "b", "c", "c", "c"],
+        "l2": ["A", "B", "C", "A", "B", "C", "A", "B", "C"],
+    })
+    assert_frame_equal(expand_grid(others=data), expected)
+
+
+def test_df_key():
+    """ Raise error if dataframe key is not supplied"""
+    df = pd.DataFrame({"x": [2, 3]})
+    others = {"df": pd.DataFrame({"x": range(1, 6), "y": [5, 4, 3, 2, 1]})}
+
+    with pytest.raises(KeyError):
+        assert expand_grid(df, others=others)
+
+
+def test_df_others():
+    """ Raise error if others is not a dict"""
+    df = pd.DataFrame({"x": [2, 3]})
+    others = [5, 4, 3, 2, 1]
+
+    with pytest.raises(TypeError):
+        assert expand_grid(df, others=others)
+
+
+def test_df_output():
+    """
+    Test output from chaining method to a dataframe.
+    Example is from tidyverse's expand_grid page - 
+    https://tidyr.tidyverse.org/reference/expand_grid.html#compared-to-expand-grid
+    """
+    df = pd.DataFrame({"x": range(1, 3), "y": [2, 1]})
+    others = {"z": range(1, 4)}
+    expected = pd.DataFrame({
+        "df_x": [1, 1, 1, 2, 2, 2],
+        "df_y": [2, 2, 2, 1, 1, 1],
+        "z": [1, 2, 3, 1, 2, 3],
+    })
+    result = expand_grid(df, df_key="df", others=others)
+    assert_frame_equal(result, expected)
+
+
+def test_df_multi_index():
+    """Test that datafarme is not a multiIndex"""
+    df = {
+        "x":
+        pd.DataFrame(
+            [[2, 3], [6, 7]],
+            columns=pd.MultiIndex.from_arrays([["m", "n"], ["p", "q"]]),
+        )
+    }
+
+    others = [5, 4, 3, 2, 1]
+
+    with pytest.raises(TypeError):
+        assert expand_grid(df, others=others)
```

Comment:
Assertion not necessary.

```suggestion
        expand_grid(df, others=others)
```